backend/shared/src/txn/run-txn.ts

```diff
@@ -1,149 +1,195 @@
 ) {
   const firestore = admin.firestore()
-  const { amount, fromType, toId, toType } = data
-  if (fromType !== 'BANK')
-    return {
-      status: 'error',
-      message: 'This method is only for transfers from banks',
-    }
+  const { amount, fromType, toId, toType, token } = data
+  if (fromType !== 'BANK') {
+    throw new APIError(400, 'This method is only for transfers from banks')
+  }
 
-  if (!isFinite(amount) || amount < 0) {
-    return { status: 'error', message: 'Invalid amount' }
+  if (!isFinite(amount) || amount <= 0) {
+    throw new APIError(400, 'Invalid amount')
   }
 
-  if (toType === 'USER') {
-    const toDoc = firestore.doc(`users/${toId}`)
-    fbTransaction.update(toDoc, {
-      balance: FieldValue.increment(amount),
-    })
-    if (!affectsProfit) {
-      fbTransaction.update(toDoc, {
-        totalDeposits: FieldValue.increment(amount),
-      })
-    }
+  if (token !== 'SPICE' && token !== 'M$') {
+    throw new APIError(400, `Invalid token type: ${token}`)
   }
 
-  const newTxnDoc = firestore.collection(`txns/`).doc()
-  const txn = {
-    id: newTxnDoc.id,
-    createdTime: Date.now(),
-    fromId: 'BANK',
-    ...data,
+  const txn = await insertTxn(pgTransaction, { fromId: 'BANK', ...data })
+
+  const update: { [key: string]: any } = {}
+
+  if (token === 'SPICE') {
+    update.spiceBalance = FieldValue.increment(amount)
+  } else {
+    update.balance = FieldValue.increment(amount)
+  }
+
+  if (!affectsProfit) {
+    update.totalDeposits = FieldValue.increment(amount)
+  }
+
+  if (toType === 'USER') {
+    await firestore.doc(`users/${toId}`).update(update)
   }
-  fbTransaction.create(newTxnDoc, removeUndefinedProps(txn))
 
-  return { status: 'success', txn }
+  return txn
 }
 
-export function runContractOldPayoutTxn(
-  fbTransaction: admin.firestore.Transaction,
-  txnData: Omit<ContractOldResolutionPayoutTxn, 'id' | 'createdTime'>
+// inserts into supabase
+export async function insertTxn(
+  pgTransaction: SupabaseTransaction,
+  txn: TxnData
 ) {
-  const firestore = admin.firestore()
-  const { amount, toId, data } = txnData
-  const { deposit } = data
-  const toDoc = firestore.doc(`users/${toId}`)
-  fbTransaction.update(toDoc, {
-    balance: FieldValue.increment(amount),
-    totalDeposits: FieldValue.increment(deposit ?? 0),
-  })
-
-  const newTxnDoc = firestore.collection(`txns/`).doc()
-  const txn = { id: newTxnDoc.id, createdTime: Date.now(), ...txnData }
-  fbTransaction.create(newTxnDoc, removeUndefinedProps(txn))
+  const row = await pgTransaction
+    .one<Row<'txns'>>(
+      `insert into txns 
+      (data, amount, from_id, to_id, from_type, to_type, category) 
+      values ($1, $2, $3, $4, $5, $6, $7) 
+      returning *`,
+      [
+        JSON.stringify(txn),
+        txn.amount,
+        txn.fromId,
+        txn.toId,
+        txn.fromType,
+        txn.toType,
+        txn.category,
+      ]
+    )
+    .catch((e) => {
+      logFailedToRecordTxn(txn)
```

Comment:
From discussion: Would be good to trigger these in dev and search the logs for them